galsim/config/value.py

```diff
@@ -618,51 +708,74 @@ def _GenerateFromSum(config, base, value_type):
     return sum, safe
- 
+
 
 def _GenerateFromCurrent(config, base, value_type):
     """@brief Get the current value of another config item.
     """
     req = { 'key' : str }
     params, safe = GetAllParams(config, base, req=req)
-
     key = params['key']
-
-    # This next bit is basically identical to the code for Dict.get(key) in catalog.py.
-    # Make a list of keys
-    chain = key.split('.')
-    d = base
-
-    # We may need to make one adjustment.  If the first item in the key is 'input', then
-    # the key is probably wrong relative to the current config dict.  We make each input
-    # item a list, so the user can have more than one input dict for example.  But if 
-    # they aren't using that, we don't want them to have to know about it if they try to 
-    # take soemthing from there for a Current item.  
-    # So we change, e.g., 
-    #     input.fits_header.file_name 
-    # --> input.fits_header.0.file_name
-    if chain[0] == 'input' and len(chain) > 2:
-        try:
-            k = int(chain[2])
-        except:
-            chain.insert(2,0)
-
-    while len(chain):
-        k = chain.pop(0)
-
-        # Try to convert to an integer:
-        try: k = int(k)
-        except ValueError: pass
-
-        if chain: 
-            # If there are more keys, just set d to the next in the chanin.
-            d = d[k]
-        else: 
-            # Otherwise, parse the value for this key
-            val,safe = ParseValue(d, k, base, value_type)
-            #print base['obj_num'],'Current key = %s, value = %s'%(key,val)
-            return val,safe
-
-    raise ValueError("Invalid key = %s given for Current"%key)
-
-
+    try:
+        return GetCurrentValue(key, base, value_type, return_safe=True)
+    except ValueError:
+        raise ValueError("Invalid key = %s given for type=Current")
+
+
+valid_value_types = {}
+
+def RegisterValueType(type_name, gen_func, valid_types):
+    """Register a value type for use by the config apparatus.
+
+    A few notes about the signature of the generating function:
+
+    1. The config parameter is the dict for the current value to be generated.  So it should
+       be the case that config['type'] == type_name.
+    2. The base parameter is the original config dict being processed.
+    3. The value_type parameter is the intended type of the generated value.  It should
+       be one of the values that you specify as valid in valid_types.
+    4. The return value of gen_func should be a tuple consisting of the value and a boolean,
+       safe, which indicates whether the generated value is safe to use again rather than
+       regenerate for subsequent objects.  This will be used upstream to determine if 
+       objects constructed using this value are safe to keep or if they have to be rebuilt.
+
+    The allowed types to include in valid_types are: float, int, bool, str, galsim.Angle,
+    galsim.Shear, galsim.PositionD.  In addition, including None in this list means that
```

Comment:
> no PositionI?

No.  We never needed one.  All the position items we build are floats, not integers.  So PositionD, not PositionI.  If a need arises, it wouldn't be hard to add this, but there hasn't been any need yet.